internal/server/server.go

```diff
@@ -108,19 +189,104 @@ func (s *Server) handleLogin() http.HandlerFunc {
 			return
 		}
 
-		session.Values["rand"] = rs
+		// generate nonce
+		nonce, err := random.SecureString(32)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for nonce")
+			return
+		}
+		session.Values["nonce"] = nonce
 
 		if err := session.Save(r, w); err != nil {
 			logAndError(w, http.StatusInternalServerError, err, "error saving session")
 			return
 		}
 
-		fmt.Fprintf(w, "session saved")
+		http.Redirect(w, r, s.oauth2Config.AuthCodeURL(randomStr, oidc.Nonce(nonce)), http.StatusSeeOther)
+	}
+}
+
+func (s *Server) handleCallback() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// handle requests that contain an error
+		if err := r.URL.Query().Get("error"); err != "" {
+			logAndError(w, http.StatusBadRequest, fmt.Errorf(err+": "+r.URL.Query().Get("error_description")), "error in request")
+			return
+		}
+
+		session := s.getSession(r)
+		oidcContext := oidc.ClientContext(r.Context(), s.client)
+
+		// handle requests without an authorization code
+		code := r.URL.Query().Get("code")
+		if code == "" {
+			logAndError(w, http.StatusUnauthorized, fmt.Errorf("authorization code empty"), "error in authorization code")
+			return
+		}
+
+		// convert authorization code into an OAuth2 token
+		oauth2Token, err := s.oauth2Config.Exchange(oidcContext, code)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error converting code to token")
+			return
+		}
+
+		// extract the ID token from OAuth2 token
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			logAndError(w, http.StatusInternalServerError, fmt.Errorf("error extracting id_token"), "no id_token in token response")
+			return
+		}
+
+		// verify the ID token
+		idToken, err := s.verifier.Verify(oidcContext, rawIDToken)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "failed to verify id_token")
+			return
+		}
+
+		// verify the OIDC nonce matches
+		if idToken.Nonce != session.Values["nonce"] {
+			logAndError(w, http.StatusUnauthorized, fmt.Errorf("error invalid nonce"), "id token nonce does not match session nonce")
+			return
+		}
+
+		// extract custom claims
+		var claims struct {
+			Email string `json:"email"`
+		}
+		if err := idToken.Claims(&claims); err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "error extracting claims")
+			return
+		}
+
+		data := map[string]interface{}{
+			"IDToken":       rawIDToken,
+			"User":          claims.Email + "_" + s.apiServerURL.Hostname(),
+			"RefreshToken":  oauth2Token.RefreshToken,
+			"APIURL":        s.apiServerURL.String(),
+			"APIHostname":   s.apiServerURL.Hostname(),
+			"ClusterCAData": s.clusterCA,
+			"IssuerURL":     s.issuerURL.String(),
+			"IssuerCAData":  s.clusterCA,
+			"ClientID":      s.kubectlClientID,
+			"ClientSecret":  s.kubectlClientSecret,
+		}
+
```

Comment:
Let's add some detailed structured logging upon generation of kubeconfig.  Include: 

- Email
- X-Forwarded-For

Reply:
Fixed